Sources/MongoSwift/MongoConnectionString.swift

```diff
@@ -0,0 +1,193 @@
+                }
+                host = String(hostAndPort[hostRange])
+                if let portRange = Range(match.range(at: 2), in: hostAndPort) {
+                    port = try HostIdentifier.parsePort(from: String(hostAndPort[portRange]))
+                }
+                self.type = .ipLiteral
+            } else {
+                let parts = hostAndPort.components(separatedBy: ":")
+                host = String(parts[0])
+                guard parts.count <= 2 else {
+                    throw MongoError.InvalidArgumentError(
+                        message: "expected only a single port delimiter ':' in \(hostAndPort)"
+                    )
+                }
+                if parts.count > 1 {
+                    port = try HostIdentifier.parsePort(from: parts[1])
+                }
+                self.type = .hostname
+            }
+            self.host = host
+            self.port = port
+        }
+
+        public var description: String {
+            var ret = ""
+            switch self.type {
+            case .ipLiteral:
+                ret += "[\(self.host)]"
+            default:
+                ret += "\(self.host)"
+            }
+            if let port = self.port {
+                ret += ":\(port)"
+            }
+            return ret
+        }
+    }
+
+    /// Parses a new `MongoConnectionString` instance from the provided string.
+    /// - Throws:
+    ///   - `MongoError.InvalidArgumentError` if the input is invalid.
+    public init(throwsIfInvalid input: String) throws {
+        let schemeAndRest = input.components(separatedBy: "://")
+        guard schemeAndRest.count == 2, let scheme = Scheme(schemeAndRest[0]) else {
+            throw MongoError.InvalidArgumentError(
+                message: "Invalid connection string scheme, expecting \'mongodb\' or \'mongodb+srv\'"
+            )
+        }
+        guard !schemeAndRest[1].isEmpty else {
+            throw MongoError.InvalidArgumentError(message: "Invalid connection string")
+        }
+        let identifiersAndOptions = schemeAndRest[1].components(separatedBy: "/")
+        // TODO: SWIFT-1174: handle unescaped slashes in unix domain sockets.
+        guard identifiersAndOptions.count <= 2 else {
+            throw MongoError.InvalidArgumentError(
+                message: "Connection string contains an unescaped slash"
+            )
+        }
+        let userAndHost = identifiersAndOptions[0].components(separatedBy: "@")
+        guard userAndHost.count <= 2 else {
+            throw MongoError.InvalidArgumentError(message: "Invalid user information")
+        }
+        let userInfo = userAndHost.count == 2 ? userAndHost[0].components(separatedBy: ":") : nil
+        let hostString = userInfo != nil ? userAndHost[1] : userAndHost[0]
+        let hosts = try hostString.components(separatedBy: ",").map(HostIdentifier.init)
+        if case .srv = scheme {
+            guard hosts.count == 1, hosts[0].port == nil else {
```

Comment:
since we're checking the port is nil below, we can remove it from this `guard`.

Reply:
done